### PR TITLE
Tic Tac Toe Plugin

### LINKED ISSUE #406: Tic Tac Toe Game
Simple Tic Tac Toe game between 2 users
Could Either 
A) Edit the original message of the board
B) Delete the old message, and send a new one on each play, just like the market place plugin.

### PR DESCRIPTION
**Pull request by tanndlin**
fixes #406

### CODE REVIEW

**Comment by tanndlin** [comment]
It's a lot, and not the most amazing code we've written
Go hard on the reviews lol

File: src/app/plugins/tictactoe.plugin.ts [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+      //If the sum of the row is 3, someone has won
+      if (Math.abs(this._sumArray(row)) === 3) {
+        flag = true;
+      }
+    });
+
+    for (let i = 0; i < 3; i++) {
+      //Get the column
+      const col = [];
+      for (let j = 0; j < 3; j++) {
```
Comment (tanndlin) [inline src/app/plugins/tictactoe.plugin.ts:206]: I don't like these kinds of for loops, but don't know a better way to do this
Lmk if someone knows a better way to get the columns

File: src/app/plugins/tictactoe.plugin.ts [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+
+      if (Math.abs(this._sumArray(col)) === 3) {
+        flag = true;
+      }
+    }
+
+    //Check diagonals
+    let sumA = 0;
+    let sumB = 0;
+    for (let i = 0; i < 3; i++) {
```
Comment (tanndlin) [inline src/app/plugins/tictactoe.plugin.ts:218]: saa

File: src/app/plugins/tictactoe.plugin.ts [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+      sumA += this._board[i][i];
+    }
+
+    for (let i = 0; i < 3; i++) {
```
Comment (tanndlin) [inline src/app/plugins/tictactoe.plugin.ts:222]: saa

**Comment by tanndlin** [comment]
@PromAtlas Still need to kill the collector when the game is over

File: src/app/plugins/tictactoe.plugin.ts [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+    }
+
+    this._board[this._col][this._row] = this.currentPlayer;
+    if (this._checkWin()) {
+      this._winner = this.currentPlayer;
+    }
+
+    this._checkTie();
+
+    this.currentPlayer *= -1;
```
Comment (c650) [inline src/app/plugins/tictactoe.plugin.ts:184]: call `flipTurn` instead

File: src/app/plugins/tictactoe.plugin.ts [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+  private flipTurn() {
+    this.currentPlayer *= -1;
+  }
+
+  private _checkWin() {
+    let flag = false;
+    this._board.forEach((row) => {
+      //If the sum of the row is 3, someone has won
+      if (Math.abs(this._sumArray(row)) === 3) {
+        flag = true;
```
Comment (c650) [inline src/app/plugins/tictactoe.plugin.ts:199]: not a fan of this side effect pattern

use a `reduce`

File: src/app/plugins/tictactoe.plugin.ts (hunk 1 above)
Comment (c650) [reply src/app/plugins/tictactoe.plugin.ts:206]: there may be an array transpose func

File: src/app/plugins/tictactoe.plugin.ts (hunk 2 above)
Comment (c650) [reply src/app/plugins/tictactoe.plugin.ts:218]: potentially some sort of filter on index

File: src/app/plugins/tictactoe.plugin.ts [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+      sumA += this._board[i][i];
+    }
+
+    for (let i = 0; i < 3; i++) {
+      sumB += this._board[i][i];
+    }
+
+    if (Math.abs(sumA) === 3 || Math.abs(sumB) === 3) {
+      flag = true;
+    }
```
Comment (c650) [inline src/app/plugins/tictactoe.plugin.ts:228]: still hate this pattern

at least, `flag = flag || Math.abs(.....`

File: src/app/plugins/tictactoe.plugin.ts [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+    this._board.forEach((row) =>
+      row.forEach((col) => {
+        if (col != 0) {
+          flag = false;
+        }
+      })
+    );
+
+    this._isTie = flag;
+  }
```
Comment (c650) [inline src/app/plugins/tictactoe.plugin.ts:249]: nest `.some` instead

**Review by c650 — COMMENTED** [review]
yucky flag pattern

**Comment by tanndlin** [comment]
fixes #406

**Pull request merged**